src/types/redis_tdigest.cc

```diff
@@ -0,0 +1,416 @@
+  auto buffer_key = internalBufferKey(ns_key, *metadata);
+  std::string buffer_value;
+  if (auto status = storage_->Get(ctx, ctx.GetReadOptions(), cf_handle_, buffer_key, &buffer_value);
+      !status.ok() && !status.IsNotFound()) {
+    return status;
+  }
+
+  for (auto item : inputs) {
+    PutDouble(&buffer_value, item);
+  }
+
+  if (auto status = batch->Put(cf_handle_, buffer_key, buffer_value); !status.ok()) {
+    return status;
+  }
+
+  return rocksdb::Status::OK();
+}
+
+rocksdb::Status TDigest::dumpCentroidsAndBuffer(engine::Context& ctx, const std::string& ns_key,
+                                                const TDigestMetadata& metadata, std::vector<Centroid>* centroids,
+                                                std::vector<double>* buffer,
+                                                ObserverOrUniquePtr<rocksdb::WriteBatchBase>* clean_after_dump_batch) {
+  if (buffer != nullptr) {
+    buffer->clear();
+    buffer->reserve(metadata.unmerged_nodes);
+    auto buffer_key = internalBufferKey(ns_key, metadata);
+    std::string buffer_value;
+    auto status = storage_->Get(ctx, ctx.GetReadOptions(), cf_handle_, buffer_key, &buffer_value);
+    if (!status.ok() && !status.IsNotFound()) {
+      return status;
+    }
+
+    if (status.ok()) {
+      rocksdb::Slice buffer_slice = buffer_value;
+      for (uint64_t i = 0; i < metadata.unmerged_nodes; ++i) {
+        double tmp_value = std::numeric_limits<double>::quiet_NaN();
+        if (!GetDouble(&buffer_slice, &tmp_value)) {
+          return rocksdb::Status::Corruption(
+              fmt::format("metadata has {} records, but get {} failed", metadata.unmerged_nodes, i));
+        }
+        buffer->emplace_back(tmp_value);
+      }
+    }
+
+    if (clean_after_dump_batch != nullptr) {
+      if (auto status = (*clean_after_dump_batch)->Delete(cf_handle_, buffer_key); !status.ok()) {
+        return status;
+      }
+    }
+  }
+
+  centroids->clear();
+  centroids->reserve(metadata.merged_nodes);
+
+  auto start_key = internalSegmentGuardPrefixKey(metadata, ns_key, SegmentType::kCentroids);
+  auto guard_key = internalSegmentGuardPrefixKey(metadata, ns_key, SegmentType::kGuardFlag);
+
+  rocksdb::ReadOptions read_options = ctx.DefaultScanOptions();
+  rocksdb::Slice upper_bound(guard_key);
+  read_options.iterate_upper_bound = &upper_bound;
+  rocksdb::Slice lower_bound(start_key);
+  read_options.iterate_lower_bound = &lower_bound;
+
+  auto iter = util::UniqueIterator(ctx, read_options, cf_handle_);
+  for (iter->SeekToFirst(); iter->Valid(); iter->Next()) {
+    Centroid centroid;
+    if (auto status = decodeCentroidFromKeyValue(iter->key(), iter->value(), &centroid); !status.ok()) {
+      return status;
+    }
+    centroids->emplace_back(centroid);
+  }
+
+  if (clean_after_dump_batch != nullptr) {
+    if (auto status = (*clean_after_dump_batch)->DeleteRange(cf_handle_, start_key, guard_key); !status.ok()) {
+      return status;
+    }
+  }
```

Comment:
why not also drop them key-by-key?

Reply:
Hi @mapleFU ,

Thanks for your suggestion! 😊
I have replaced it with key-by-key deletion.

Best Regards,
Edward